components/search/backend/src/main/kotlin/SearchService.kt

```diff
@@ -70,50 +81,133 @@ class SearchService(
         )
 
         return db.blockingQuery {
-            val query = OrtRunsTable
-                .innerJoin(AnalyzerJobsTable, { OrtRunsTable.id }, { ortRunId })
-                .innerJoin(AnalyzerRunsTable, { AnalyzerJobsTable.id }, { analyzerJobId })
-                .innerJoin(PackagesAnalyzerRunsTable, { AnalyzerRunsTable.id }, { analyzerRunId })
-                .innerJoin(PackagesTable, { PackagesAnalyzerRunsTable.packageId }, { PackagesTable.id })
-                .innerJoin(IdentifiersTable, { PackagesTable.identifierId }, { IdentifiersTable.id })
-                .innerJoin(RepositoriesTable, { OrtRunsTable.repositoryId }, { RepositoriesTable.id })
-                .innerJoin(ProductsTable, { RepositoriesTable.productId }, { ProductsTable.id })
-
-            val concatenatedIdentifier = concat(
-                IdentifiersTable.type,
-                stringLiteral(":"),
-                IdentifiersTable.namespace,
-                stringLiteral(":"),
-                IdentifiersTable.name,
-                stringLiteral(":"),
-                IdentifiersTable.version
-            )
-
-            val identifierCondition = concatenatedIdentifier.applyRegex(identifier)
+            val query = createBaseQuery()
 
-            val whereClause = hierarchyFilter.apply(identifierCondition) { level, ids, filter ->
-                generateHierarchyCondition(level, ids, filter)
+            when {
+                purl != null -> findByPurl(query, purl, hierarchyFilter)
+                identifier != null -> findByIdentifier(query, identifier, hierarchyFilter)
+                else -> emptyList()
             }
+        }
+    }
+
+    private fun createBaseQuery() = OrtRunsTable
+        .innerJoin(AnalyzerJobsTable, { OrtRunsTable.id }, { ortRunId })
+        .innerJoin(AnalyzerRunsTable, { AnalyzerJobsTable.id }, { analyzerJobId })
+        .innerJoin(PackagesAnalyzerRunsTable, { AnalyzerRunsTable.id }, { analyzerRunId })
+        .innerJoin(PackagesTable, { PackagesAnalyzerRunsTable.packageId }, { PackagesTable.id })
+        .innerJoin(IdentifiersTable, { PackagesTable.identifierId }, { IdentifiersTable.id })
+        .innerJoin(RepositoriesTable, { OrtRunsTable.repositoryId }, { RepositoriesTable.id })
+        .innerJoin(ProductsTable, { RepositoriesTable.productId }, { ProductsTable.id })
+
+    /**
+     * Search by PURL using curated PURL resolution.
+     * The effective PURL is determined by COALESCE(curated_purl, original_purl).
+     */
+    private fun findByPurl(
+        query: org.jetbrains.exposed.sql.Join,
+        purl: String,
+        hierarchyFilter: HierarchyFilter
+    ): List<RunWithPackage> {
+        val effectivePurl = createEffectivePurlExpression()
+
+        @Suppress("UNCHECKED_CAST")
+        val purlCondition = (effectivePurl as Expression<String>).applyRegex(purl)
```

Comment:
Unchecked casts are always suspicious. I haven't checked, but please verify that this is really needed.